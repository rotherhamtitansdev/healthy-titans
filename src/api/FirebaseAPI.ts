import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
