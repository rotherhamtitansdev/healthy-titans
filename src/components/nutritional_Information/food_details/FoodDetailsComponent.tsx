import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import NutritionBreakdownChart from "./nutrition_breakdown_chart/NutritionBreakdownChart";
import FoodDetailsComponentData, {
  FoodDetailsProps,
} from "../../../data/nutritional_information/FoodDetailsComponentData";
import DetailsCard from "../../shared/DetailsCard";
import FirebaseAPI from "../../../api/FirebaseAPI";
import DetailsComponent from "../../shared/DetailsComponent";

/* eslint-disable */

const FoodDetailsComponent = () => {
  const [getFoodDetailsComponentData, setFoodDetailsComponentData] = useState<
    FoodDetailsProps | undefined
  >();

  const [getImageURL, setImageURL] = useState<string>();
  const { foodName } = useParams();

  useEffect(() => {
    // @ts-ignore
    const data = FoodDetailsComponentData[foodName];
    if (data.firebaseName) {
      FirebaseAPI.fetchImages(data.firebaseName).then((res) =>
        setImageURL(res)
      );
    } else if (data.img) {
      setImageURL(data.img);
    }

    setFoodDetailsComponentData(data);
  }, []);

  return (
    <div>
      {getFoodDetailsComponentData ? (
        <DetailsComponent>
          <DetailsCard
            name={getFoodDetailsComponentData.name}
            description={getFoodDetailsComponentData.description}
            img={getImageURL}
            additionalStyling="lg:w-5/12"
            list={
              getFoodDetailsComponentData.facts && {
                title: "More Facts",
                items: getFoodDetailsComponentData.facts,
              }
            }
          />
          <NutritionBreakdownChart name={getFoodDetailsComponentData.name} />
        </DetailsComponent>
      ) : (
        <h1>loading</h1>
      )}
    </div>
  );
};

export default FoodDetailsComponent;
