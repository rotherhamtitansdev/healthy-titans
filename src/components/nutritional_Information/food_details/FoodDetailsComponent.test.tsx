import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router";
import FirebaseAPI from "../../../api/FirebaseAPI";
import FoodDetailsComponent from "./FoodDetailsComponent";

const mockContent = {
  cardData: [
    {
      category: "mockCategory",
      description: "mockDescription",
      facts: ["mockFact1", "mockFact2"],
      firebaseName: "mock/Image.svg",
      name: "mock1",
      score: 0,
    },
    {
      category: "mockCategory",
      description: "mockDescription2",
      facts: ["mockFact12", "mockFact22"],
      firebaseName: "mock/Image2.svg",
      name: "mock2",
      score: 2,
    },
    {
      category: "mockCategory",
      description: "mockDescription3",
      facts: ["mockFact13", "mockFact23"],
      firebaseName: "mock/Image3.svg",
      name: "mock3",
      score: 9,
    },
  ],
  paths: ["mock1", "mock2", "mock3"],
};

describe("Food Details Component", () => {
  test("component renders both correct data and image from path and a carousel", async () => {
    jest.spyOn(FirebaseAPI, "fetchFoodDetailsSeeNext").mockResolvedValue(mockContent);
    jest.spyOn(FirebaseAPI, "fetchFoodDetailsSingle").mockResolvedValue(mockContent.cardData[2]);
    jest.spyOn(FirebaseAPI, "fetchImages").mockResolvedValue("mock/Image3.svg");

    // define width for carousel to appear
    Object.defineProperties(window.HTMLElement.prototype, {
      offsetWidth: {
        get: () => 1400,
      },
    });

    // Using memoryrouter to mock Route as path to the component is required for the contents
    const { asFragment } = render(
      <MemoryRouter
        initialEntries={[{ pathname: "/NutritionalInformation/mockCategory/mock3" }]}
        initialIndex={0}
      >
        <Routes>
          <Route
            path="/NutritionalInformation/mockCategory/:foodName"
            element={<FoodDetailsComponent />}
          />
        </Routes>
      </MemoryRouter>
    );
    const mockTextToFind = await screen.findAllByText("mock3");
    const mockButtonToFind = await screen.findAllByRole("button");

    expect(mockTextToFind[0]).toBeVisible();
    expect(await screen.findByText("mockFact23")).toBeVisible();
    expect(await screen.findByRole("img", { name: "mock3" })).toHaveAttribute(
      "src",
      "mock/Image3.svg"
    );
    expect(mockButtonToFind[0]).toHaveAttribute("data-testid", "mock1");
    expect(asFragment()).toMatchSnapshot();
  });
});
