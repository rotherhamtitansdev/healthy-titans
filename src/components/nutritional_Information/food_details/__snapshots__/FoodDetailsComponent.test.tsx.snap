// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Food Details Component component renders both correct data and image from path and a carousel 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="px-10 lg:px-6 pt-4 lg:container mx-auto"
    >
      <div
        class="text-[#243A73] flex font-bold sm:text-[18px] pb-10"
        data-testid="breadcrumbs"
      >
        <a
          aria-current="page"
          class="active"
          href="/NutritionalInformation"
        >
           Food &  Nutrition
        </a>
        <p
          class="text text-titansBrightPink"
        >
           &gt; 
        </p>
        <a
          aria-current="page"
          class="active"
          href="/NutritionalInformation/mockCategory"
        >
           Mock Category
        </a>
        <p
          class="text text-titansBrightPink"
        >
           &gt; 
        </p>
        <a
          aria-current="page"
          class="active"
          href="/NutritionalInformation/mockCategory/mock3"
        >
           Mock3
        </a>
      </div>
      <div
        class="bg-white rounded-2xl flex flex-wrap md:flex-nowrap gap-y-5 gap-x-0 md:gap-x-2 mt-[6rem] mb-[6rem]"
      >
        <div
          class="pl-5 md:basis-1/2"
          data-testid="mock3-details"
        >
          <img
            alt="mock3"
            class="absolute visible top-[13rem]"
            src="mock/Image3.svg"
          />
          <div
            class="pl-2 pr-4 pt-[5rem]"
          >
            <div
              class="tracking-wide text-4xl text-black font-bold py-5 text-homepageHeaderSubTitle"
              data-testid="details-title"
            >
              mock3
            </div>
            <p
              class="block mt-1 text-xl leading-tight font-medium text-gray-900 text-[16px] lg:text-[20px]"
              data-testid="details-description"
            >
              mockDescription3
            </p>
            <div
              class="block mt-1 text-lg leading-tight font-semibold text-gray-900"
            >
              <h2
                class="py-5 text-[20px] lg:text-[24px] "
              >
                More Facts
              </h2>
              <ul
                class="list-none px-0 lg:px-5 text-[16px] lg:text-[20px]"
              >
                <li
                  class="pb-4 md:pb-8 flex font-medium"
                >
                  <img
                    alt="checkmark"
                    class="pr-5"
                    src="CheckMarkButton.svg"
                  />
                  mockFact13
                </li>
                <li
                  class="pb-4 md:pb-8 flex font-medium"
                >
                  <img
                    alt="checkmark"
                    class="pr-5"
                    src="CheckMarkButton.svg"
                  />
                  mockFact23
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="md:basis-1/2"
        >
          <div
            class="bg-white rounded-3xl w-fit md:w-full lg:w-full xl:w-[40rem] mx-auto"
          >
            <span
              style="width: 150px; height: 150px; border-radius: 100%; border: 2px solid; border-top-color: red; border-bottom-color: transparent; border-left-color: red; border-right-color: red; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="-mt-10 w-full lg:container mx-auto px-2 sm:px-6 lg:px-6 pb-20"
    >
      <p
        class="text-titansDarkBlue text-[24px] sm:text-[36px] pl-4 sm:pl-0 font-semibold pb-6"
      >
        See next
      </p>
      <div
        class="react-multi-carousel-list  "
        dir="ltr"
      >
        <ul
          class="react-multi-carousel-track "
          style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
        >
          <li
            aria-hidden="false"
            class="react-multi-carousel-item react-multi-carousel-item--active p-[1rem]"
            data-index="0"
            style="flex-basis: auto; position: relative; width: 467px;"
          >
            <div
              class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
              data-testid="mock1"
              role="button"
              tabindex="0"
            >
              <div
                class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
              >
                <div
                  class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
                >
                  <div
                    class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
                  >
                    <img
                      alt="mock1"
                      class="mx-auto undefined"
                      draggable="false"
                      src="mock/Image3.svg"
                    />
                  </div>
                </div>
                <div
                  class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
                >
                  mock1
                </div>
              </div>
            </div>
          </li>
          <li
            aria-hidden="false"
            class="react-multi-carousel-item react-multi-carousel-item--active p-[1rem]"
            data-index="1"
            style="flex-basis: auto; position: relative; width: 467px;"
          >
            <div
              class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
              data-testid="mock2"
              role="button"
              tabindex="0"
            >
              <div
                class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
              >
                <div
                  class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
                >
                  <div
                    class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
                  >
                    <img
                      alt="mock2"
                      class="mx-auto undefined"
                      draggable="false"
                      src="mock/Image3.svg"
                    />
                  </div>
                </div>
                <div
                  class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
                >
                  mock2
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
